Fix: ensure input line does not lose focus on switching between profiles

SlySven opened the pull request:
This should close #6485.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: ensure input line does not lose focus on switching between profiles

This should close #6485.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

wiploo commented:
Sorry, but nothing change for me.

Focus is always on the toolbar.

Thanks for this PR

SlySven committed:
Revise: fix focus being directed to the main toolbar

For some inexplicable reason the code to switch between multiple loaded
profiles seemed to redirect the focus to buttons on the main toolbar that
have drop-down menus - instead of the wanted command-line - though there is
no indication of an explicit focus policy to do this (maybe it is a built-
in one); this seems to be sufficient to counter it.

There is one down-side in that it is not possible to navigate to the main
toolbar now with `<CTRL>+<TAB>` - it may be desirable to include it in the
sequence that that key-press combination processes.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: change the focus modification of some main toolbar buttons

Careful studying of the current `development` code with GammaRay revealed
that the ordinary buttons on the Main Toolbar had the `Qt::NoFocus` policy
and that the ones with a drop down menu had `Qt::TabFocus` - this suggests
that I should have changed the policy in the prior commit for the latter
to also be `Qt::NoFocus` as well (instead of `Qt::ClickFocus`).

Note: it would still be desirable to allow the end-user to `<CTRL>+<TAB>`/
`<CTRL>+<BACKTAB>` to the main toolbar - which I think would require
enabling `Qt::TabFocus` on ALL the main toolbar buttons when the first or
last profile is the active one AND also arrange for it to be be disabled/
overridden when the appropriate key-event happens on the end two buttons
on the Toolbar - however the latter will require coding an event-handler
for that toolbar which is too much to do here and should be considered as
a separate issue.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

wiploo commented:
Unfortunately the problem is present even in the last build. The toolbar is not focus any more, but the main input is not focused.

Thanks for this PR.

SlySven commented:
> Unfortunately the problem is present even in the last build. The toolbar is not focus any more, but the main input is not focused.

Can you tell where it ***is*** going to? Is it like that all the time or, if you click on the main input does it then behave sensibly afterwards? I ask this because I do not understand whether I need to fix something that only happens after profiles is/are opened or all the time...

SlySven committed:
Revise: record which command line is being used for each profile

So that we can return the focus to it when we switch between profiles.

This is an amalgamation with the code from #6232 and recent changes in the
main `development` branch. However because the development of this PR
has taken so long it has been difficult to clean it up whilst also merging
in the other changes so I have taken the step of saving copies of all the
files that my working code ends up as and then injected them into the last
published commit. This might seem messy but it was the only way I could see
to keep the bits I wanted but not the bits I didn't.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Still working on this - have made progress but not done yet...

Kebap commented:
@SlySven asked:
> Can you tell where it _**is**_ going to? Is it like that all the time or, if you click on the main input does it then behave sensibly afterwards? I ask this because I do not understand whether I need to fix something that only happens after profiles is/are opened or all the time...

Clicking the input lines works fine still. Only it should not be necessary. Simple CTRL+TAB should (and did) do the trick.
Right now with the current fix, the focus is not going anywhere! You can easily review if you enable multi-view:
Open two profiles A and B. Click in main input of A. Press CTRL-Tab. Input stays where it was. Press CTRL-Tab again. Now you should (and indeed still are) back in A. 
It is the same without multi-view. Only then profile B is displayed alright, so you don't see the focus staying in input A immediately.

SlySven commented:
The trouble I am finding is that, as far as I can make out `<Ctrl>+<Tab>` is *also* used/needed to navigate around the widgets in a window when `<Tab>` on its own is not enough.

I am still working on this though to see if I can make things better - it is very tricky to debug though, because any interaction with the Qt Creator debugger also switches focus away from the Mudlet application which immediately invalidates the situation - so instead I have to plaster the code-base with `qDebug()` lines and try and track what is happening...

:hammer_and_wrench: 

Also, the example you quote is only one "simple" case, when there are:
* additional `TCommandLine`s in the main console,
* mini-consoles anchored into the main console - some with their command line enabled,
* floating/dockable widgets (mapper / user windows) - the latter maybe with their own command line or additional ones on top
* TLabels
* floating/dockable Toolbars
* mouse-clicks on any of the above

there are a lot more cases to get the focus movement right for.

Kebap commented:
> see if I can make things better

Please just restore the previous functionality for now. 
Remember there is a feature freeze and upcoming release..

SlySven commented:
> > see if I can make things better
> 
> Please just restore the previous functionality for now. Remember there is a feature freeze and upcoming release..

Well - how far back do you want me to revert? Give me a SHA1 and I'll put it back and try to get it sorted for a **4.17.1** - I am not looking at anything else at present. :cry:

Kebap commented:
bcross on Discord added, the Ctrl + Tab (and Ctrl + 1, etc.) most probably still worked before PR #6483 asked to review screenreader functionality.

Kebap commented:
> most probably still worked before PR #6483

Git bisect found the first broken version was #6267 from 19.12.2022
The version before this was #6458 and still worked fine as expected.

SlySven commented:
Sadly, that #6267 broke things is not surprising given that it was supposed to organise the switching between profiles better/more cleanly than before. I will continuing to focus on it once I've done #6232 - ironically the two of them will interact with each other so the sooner I can close that the better.

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

SlySven commented:
This does include the code from #6232 (as it is currently) so the changes will seem larger and hit more files **until** the other PR gets merged in - any further changes *there* will likely need to be imported here. OTOH if this one does what it promises it will finally sort out something I have been stuck with for the last couple of weeks. :weary:

SlySven committed:
Update: pull in changes make to #6232

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

Conflicts resolved in:
* src/Host.h
* src/TConsole.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

SlySven committed:
Fix: remove many setFocusPolicy(...) statements that are probably unneeded

I found out that setting a focusPolicy on a widget that itself has a
focusProxy (so that it sends the focus on it to that proxy) has the effect
of applying the policy to the *proxy*. Unfortunately that meant that
somehow with the previous commits to this PR, it was possible for the
main command line to end up with a "NoFocus" policy applied to it with the
result that after creating or enabling other `TCommandLine` instances for
a profile it became *impossible* to click on the main command line so that
it got the focus and was able to receive keyboard input.

As the default focusPolicy for *most* widgets is `NoFocus` anyhow I felt it
was probably safe to remove any policy settings, particular to that type,
so it would not get sent/set accidentally on the main command line when it
was the focusProxy for those widgets. The cases where it might be required,
i.e. any `TCommandLine`, are set to `StrongFocus` in the constructor.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

Needed to pull in fixes for failing MacOS CI builds (#6552).

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

wiploo commented:
I have just tested it and it works very well. Thanks @SlySven!

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

atari2600tim commented:
The code is over my head, but in usage it is improved just as I had hoped.

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

SlySven committed:
Merge branch 'development' into Fix_ensureInputLineDoesNotLoseFocusOnSwitchingBetweenProfiles

vadi2 commented on src/TConsole.cpp:
```diff
@@ -2203,53 +2240,21 @@ void TConsole::setCaretMode(bool enabled)
+        Q_ASSERT_X(!mUpperPane->focusProxy(), "TConsole:setCaretMode(false) FAIL", "About to set a focusPolicy but there is a focusProxy in place that will get it instead!");
+        mUpperPane->setFocusPolicy(Qt::ClickFocus);
+        setProxyForFocus(mpCommandLine);
+        // Carefull - if there is a FocusProxy for this element then IT gets the policy
```
```suggestion
        // Careful - if there is a FocusProxy for this element then IT gets the policy
```

vadi2 commented on src/Host.cpp:
```diff
@@ -4076,11 +4076,55 @@ void Host::setCaretEnabled(bool enabled) {
+        if (!pCommandLine) {
+            mpLastCommandLineUsed.pop();
+        }
+    } while (!mpLastCommandLineUsed.isEmpty() && !pCommandLine);
```
This doesn't seem safe, what if the stack goes empty and no command line was found?

SlySven replied:
It'll end up returning a `nullptr` which the caller will have to deal with...

vadi2 commented on src/TCommandLine.cpp:
```diff
@@ -177,13 +177,10 @@ bool TCommandLine::event(QEvent* event)
-                } else {
-                    mudlet::self()->slot_tabChanged(currentIndex - 1);
-                }
+                const int currentIndex = mudlet::self()->mpTabBar->currentIndex();
+                const int count = mudlet::self()->mpTabBar->count();
```
👍

vadi2 commented on src/TDockWidget.h:
```diff
@@ -33,14 +34,16 @@
 
 // TDockWidget contains helpers for User Windows QDockWidget.
 class TDockWidget : public QDockWidget {
+    Q_OBJECT // Needed so we can use qobject_cast on it!
```
```suggestion
    Q_OBJECT
```

SlySven replied:
I wasn't sure that it was obvious - hence the comment.

vadi2 replied:
I don't think it needs to be explained, we're using Qt after all and this is a common macro
